<a href="https://colab.research.google.com/github/JaySanthanam/Programming-for-data/blob/main/Worksheets/Data_Exploration_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exploring data sets
---

This [link](https://docs.google.com/document/d/1cijDOCDixsYu-Rr9pC8STPPXado3xoFpgBAZgdDTLHs/edit?usp=sharing) will take you to a list of some of the datasets we use during the Programming for Data course.

For this exercise, please choose a dataset to work with.

Use the code cell below to play with the data.  Look at what data is included, where there is any null data, if there are any columns that need renaming, removing, new values calculated from them.

Form some ideas of what you might want to know from the data and use the techniques you have learned so far to show the answers to these.  You might do this by getting summary statistics, by sorting and showing parts of the dataset, by calculating new values, etc.

### What to do
---

Decide what you want to know, explain this in the TEXT cell below, then write some code to get the answer.  When you have done this, add a new text cell below the code cell, explain what you have found.  Then do the same again, adding text cell, code cell, text cell, for something else you would like to know.

I have chosen Exam daat to work with as an example.

First thing I want to do is read the data from the link and look at the headings and the first few rows to know what the data looks like and also know the shape of my dataframe. I shall call this part as data retrieval (don't know what the right word for this is). Below is the code for this.

In [5]:
def get_data_retrieval():
  url =  "https://raw.githubusercontent.com/lilaceri/Working-with-data-/main/Data%20Sets%20for%20code%20divisio/exams.csv"
  explore_data = pd.read_csv(url)
  return explore_data


import pandas as pd

data_explore = get_data_retrieval()
print(data_explore.shape)
print(data_explore.head())
print(data_explore.columns)

(1000, 8)
   gender ethnicity  ... reading_score writing_score
0  female   group E  ...            76            71
1    male   group C  ...            72            69
2  female   group B  ...            56            61
3    male   group D  ...            60            59
4  female   group C  ...            54            55

[5 rows x 8 columns]
Index(['gender', 'ethnicity', 'parental_education', 'lunch',
       'preparation_course', 'math_score', 'reading_score', 'writing_score'],
      dtype='object')


I now know that I have 1000 rows and 8 columns in the dataframe
The 8 column headings are 'gender', 'ethnicity', 'parental_education', 'lunch', 'preparation_course', 'math_score', 'reading_score' and 'writing_score'. They are all saved as objects datatype. There are no time or date variables (so I will not need to import datetime libraries). Now it's time to find any null values. Also, I am going to drop "preparation_course" columns. Since I will not be using it further in my analyses.

In [57]:
def any_nulls(df):
  df_null = df.isna().any()
  return df_null

data_explore_null = any_nulls(data_explore)
print(data_explore_null)

# drop column on preparatory course since it doesn't matter for my further analysis - I'm not going to use it.
data_explore = data_explore.drop(columns=['preparation_course'])
print(data_explore.columns)


gender                False
ethnicity             False
parental_education    False
lunch                 False
math_score            False
reading_score         False
writing_score         False
dtype: bool
Index(['gender', 'ethnicity', 'parental_education', 'lunch', 'math_score',
       'reading_score', 'writing_score'],
      dtype='object')


There are no nulls in the dataframe. So now, I want to look at the unique values in the columns to understand what the columns denote.

In [58]:
df = data_explore
for col in df:
  print(df[col].unique())

['female' 'male']
['group E' 'group C' 'group B' 'group D' 'group A']
['college' 'high school' "associate's degree" "master's degree"
 "bachelor's degree"]
['free/reduced' 'standard']
[ 65  75  62  60  34  53  85  93  58  83  54  38  90  61  55  44  71  77
  39  78  74  56  68  48  82  66  64  89  87  69  50  80  99  57  59  96
  91  41  52  98  51  79  84  92  37  81  49  21  95  73  70  47  63  43
  46  76  72  29  42  67 100  40  33  86  36  30  94  35  45  88  18  31
  97  23  28  32  27  13]
[ 76  72  56  60  54  66  88  73  71  75  94  69  36  81  57  64  79  35
  85  82  63  39  92  80  68  83  89 100  77  59  84  99  97  67  65  87
  70  43  44  91  46  74  90  26  52  48  61  62  55  41  53  78  58  32
  98  42  30  27  47  51  49  40  45  28  86  93  17  50  38  37  95  16]
[ 71  69  61  59  55  92  70  63  82  42  81  52  62  73  56  49  72  33
  86  65  43  84  83  54  79  80  99  75  85  87 100  74  77  93  57  67
  64  68  41  60  38  20  53  45  78  76  90  48  44  66  3

Now that I see unique values of the columns, I want to replace "some college" to college and "some high school" to "high school" in column parental education (It doesn't make sense to me about "some college or highschool".

In [15]:
def clean_data(df):
  df["parental_education"] = df["parental_education"].replace('some college','college')
  df["parental_education"] = df["parental_education"].replace('some high school','high school')
  return df

data_explore_clean = clean_data(data_explore)

So now that I have cleaned up the data and have an idea of the dataframe. I want to understand the following:
Which gender has the highest 
1. 'math_score' 
2. 'reading_score' 
3. 'writing_score'

respectively.

In [59]:
def math_gender_describe(df):
  math_gender = df.groupby('gender').math_score. describe()
  return math_gender

def write_gender_describe(df):
  write_gender = df.groupby('gender').writing_score. describe()
  return write_gender

def read_gender_describe(df):
  read_gender = df.groupby('gender').reading_score. describe()
  return read_gender

print("Math scores by gender\n")
gender_math = math_gender_describe(data_explore)
print(gender_math)

print("\nReading scores by gender\n")
gender_read = read_gender_describe(data_explore)
print(gender_read)

print("\nWriting scores by gender\n")
gender_write = write_gender_describe(data_explore)
print(gender_write)


Math scores by gender

        count       mean        std   min    25%   50%   75%    max
gender                                                             
female  496.0  64.941532  14.544623  18.0  54.75  66.0  75.0  100.0
male    504.0  68.748016  15.642438  13.0  59.00  69.0  79.0  100.0

Reading scores by gender

        count       mean        std   min    25%   50%   75%    max
gender                                                             
female  496.0  73.866935  13.867717  17.0  64.00  75.0  84.0  100.0
male    504.0  65.648810  14.537320  16.0  55.75  66.0  75.0  100.0

Writing scores by gender

        count       mean        std   min   25%   50%   75%    max
gender                                                            
female  496.0  73.651210  14.030606  17.0  64.0  75.0  83.0  100.0
male    504.0  63.581349  15.125497  12.0  53.0  64.0  74.0  100.0


Naturally, we can draw conclusions based on the above statistics. There are more males than females (but not significantly different). In reading and writing females (on an average) score better than males. In maths males score better on an average than females. In maths, reading and writing the highest scores are the same between males and females( 100%). However, the lowest scores are worse among male than female in all 3 subjects. Of course, I can make more inferences if I look at percentiles etc. But I will stop here.

Next, I want to know the impact of ethinic background on the exam scores. Again, this can be done in a similar way to gender. Before, I group by both gender and ethinic background, let's look at just ethnic background impact on it's own on the average score in all 3 subjects.

In [60]:
def math_ethnic_describe(df):
  math_ethnic = df.groupby('ethnicity').math_score. describe()
  return math_ethnic

def write_ethnic_describe(df):
  write_ethnic = df.groupby('ethnicity').writing_score. describe()
  return write_ethnic

def read_ethnic_describe(df):
  read_ethnic = df.groupby('ethnicity').reading_score. describe()
  return read_ethnic

print("Math scores by ethnicity\n")
ethnic_math = math_ethnic_describe(data_explore)
print(ethnic_math)

print("\nReading scores by ethnicity\n")
ethnic_read = read_ethnic_describe(data_explore)
print(ethnic_read)

print("\nWriting scores by ethnicity\n")
ethnic_write = write_ethnic_describe(data_explore)
print(ethnic_write)

Math scores by ethnicity

           count       mean        std   min   25%   50%    75%    max
ethnicity                                                             
group A     82.0  64.378049  15.144278  23.0  57.0  65.0  75.75  100.0
group B    197.0  62.832487  14.246638  28.0  53.0  63.0  73.00  100.0
group C    313.0  64.188498  15.231323  13.0  54.0  66.0  74.00  100.0
group D    254.0  68.433071  14.549678  18.0  59.0  67.0  79.00  100.0
group E    154.0  76.168831  13.335198  45.0  69.0  77.0  84.00  100.0

Reading scores by ethnicity

           count       mean        std   min   25%   50%    75%    max
ethnicity                                                             
group A     82.0  68.926829  14.729479  30.0  59.0  70.0  80.75   97.0
group B    197.0  66.487310  14.533076  36.0  56.0  65.0  78.00  100.0
group C    313.0  68.217252  15.112529  16.0  58.0  69.0  79.00  100.0
group D    254.0  70.889764  14.314772  17.0  61.0  72.0  80.00  100.0
group E    154.0  75.

Group C has the highest number of pupils in the course. Group A has lowest number of pupils taking the course. Group E pupils on an average have scored higher in all 3 subjects. The lowest scores in all 3 subjects are from Group C students.

Now, let's try and explore how the gender and ethnicity play out together by grouping by both gender and ethnicity.

In [61]:
def math_ethnic_gender(df):
  math_ethnic_gender = df.groupby(['ethnicity','gender']).math_score. describe()
  return math_ethnic_gender

def write_ethnic_gender(df):
  write_ethnic_gender = df.groupby(['ethnicity','gender']).writing_score. describe()
  return write_ethnic_gender

def read_ethnic_gender(df):
  read_ethnic_gender = df.groupby(['ethnicity','gender']).reading_score. describe()
  return read_ethnic_gender

print("Math scores by ethnicity and Gender\n")
ethnic_gender_math = math_ethnic_gender(data_explore)
print(ethnic_gender_math)

print("\nReading scores by ethnicity and Gender\n")
ethnic_gender_read = read_ethnic_gender(data_explore)
print(ethnic_gender_read)

print("\nWriting scores by ethnicity and Gender\n")
ethnic_gender_write = write_ethnic_gender(data_explore)
print(ethnic_gender_write)

Math scores by ethnicity and Gender

                  count       mean        std   min    25%   50%    75%    max
ethnicity gender                                                              
group A   female   38.0  63.736842  13.865540  31.0  57.00  64.5  75.75   86.0
          male     44.0  64.931818  16.307222  23.0  56.75  65.5  75.25  100.0
group B   female   93.0  60.666667  14.335524  28.0  52.00  61.0  73.00   89.0
          male    104.0  64.769231  13.951532  35.0  55.00  65.0  73.25  100.0
group C   female  163.0  63.079755  13.842581  27.0  51.50  65.0  72.00   94.0
          male    150.0  65.393333  16.572354  13.0  55.00  66.0  76.00  100.0
group D   female  123.0  65.105691  14.280454  18.0  56.00  65.0  73.50   93.0
          male    131.0  71.557252  14.152554  31.0  62.50  70.0  82.00  100.0
group E   female   79.0  74.139241  13.285556  45.0  67.50  74.0  82.50  100.0
          male     75.0  78.306667  13.137499  47.0  71.00  79.0  87.00  100.0

Reading scores

The number of female and male students is pretty similar in all ethinic groups (except groups B and C that have the highest difference in numbers- 11 and 13 respectively). Regardless of ethinic groups, females score better on an average in both reading and writing while they score lower than males on an average in maths. Group E females score better than everyone in reading and writing and group E males score better than everyone in maths. Group D females and Group C males have the lowest scores in all 3 subjects.

Next I want to look at the impact of financial challenges on pupil's scores. To do this, I am going to group by lunch (this is based on the assumption that those who get free/reduced lunch options are facing financial challenges). I would also like to see how teh financial challenge impacts males and females pupils and also pupils from different ethinicity. The code for this is below.

In [63]:
def math_lunch_describe(df):
  math_lunch = df.groupby('lunch').math_score. describe()
  return math_lunch

def write_lunch_describe(df):
  write_lunch = df.groupby('lunch').writing_score. describe()
  return write_lunch

def read_lunch_describe(df):
  read_lunch = df.groupby('lunch').reading_score. describe()
  return read_lunch

print("Math scores by Financial difficulty\n")
lunch_math = math_lunch_describe(data_explore)
print(lunch_math)

print("\nReading scores by Financial difficulty\n")
lunch_read = read_lunch_describe(data_explore)
print(lunch_read)

print("\nWriting scores by Financial Difficulaty\n")
lunch_write = write_lunch_describe(data_explore)
print(lunch_write)


def math_lunch_gender(df):
  math_lunch_gender = df.groupby(['lunch','gender']).math_score. describe()
  return math_lunch_gender

def write_lunch_gender(df):
  write_lunch_gender = df.groupby(['lunch','gender']).writing_score. describe()
  return write_lunch_gender

def read_lunch_gender(df):
  read_lunch_gender = df.groupby(['lunch','gender']).reading_score. describe()
  return read_lunch_gender

print("\nMath scores by Financial difficulty and Gender\n")
gender_lunch_math = math_lunch_gender(data_explore)
print(gender_lunch_math)

print("\nReading scores by Financial difficulty and Gender\n")
gender_lunch_read = read_lunch_gender(data_explore)
print(gender_lunch_read)

print("\nWriting scores by Financial difficulty and Gender\n")
gender_lunch_write = write_lunch_gender(data_explore)
print(gender_lunch_write)

def math_lunch_ethnic(df):
  math_lunch_ethnic = df.groupby(['lunch','ethnicity']).math_score. describe()
  return math_lunch_ethnic

def write_lunch_ethnic(df):
  write_lunch_ethnic = df.groupby(['lunch','ethnicity']).writing_score. describe()
  return write_lunch_ethnic

def read_lunch_ethnic(df):
  read_lunch_ethnic = df.groupby(['lunch','ethnicity']).reading_score. describe()
  return read_lunch_ethnic

print("\nMath scores by Financial difficulty and ethnicity\n")
ethnic_lunch_math = math_lunch_ethnic(data_explore)
print(ethnic_lunch_math)

print("\nReading scores by Financial difficulty and ethnicity\n")
ethnic_lunch_read = read_lunch_ethnic(data_explore)
print(ethnic_lunch_read)

print("\nWriting scores by Financial difficulty and ethnicity\n")
ethnic_lunch_write = write_lunch_ethnic(data_explore)
print(ethnic_lunch_write)

Math scores by Financial difficulty

              count       mean        std   min   25%   50%   75%    max
lunch                                                                   
free/reduced  353.0  59.181303  13.961635  13.0  49.0  60.0  69.0   93.0
standard      647.0  71.049459  14.220206  18.0  62.0  71.0  81.0  100.0

Reading scores by Financial difficulty

              count       mean        std   min   25%   50%   75%    max
lunch                                                                   
free/reduced  353.0  64.776204  14.022437  16.0  56.0  65.0  74.0   97.0
standard      647.0  72.425039  14.499072  17.0  63.0  73.0  83.0  100.0

Writing scores by Financial Difficulaty

              count       mean        std   min   25%   50%   75%    max
lunch                                                                   
free/reduced  353.0  63.062323  14.876827  12.0  53.0  64.0  74.0  100.0
standard      647.0  71.584235  14.897958  17.0  62.0  73.0  82.0  100.0

Mat

More than 1/3 of the pupils are facing financial difficulties. Financial difficulties has a huge impact on the average scores of the pupils in all 3 subjects. It has slightly larger impact on the male students compared to females. Group E has the lowest number of pupils facing financial difficulties and naturally pupils from group E seem to do better in all 3 subjects. 

Next I want to look at the impact of parental education in the pupils scores in all 3 subjects. I will then continue to look at any differences in gender or ethinicity results based on parental education. To do this, first I will group by parental education. Then by parental education and gender and then by parental education and ethinicity.

In [64]:
def math_parents_describe(df):
  math_parents = df.groupby('parental_education').math_score. describe()
  return math_parents

def write_parents_describe(df):
  write_parents = df.groupby('parental_education').writing_score. describe()
  return write_parents

def read_parents_describe(df):
  read_parents = df.groupby('parental_education').reading_score. describe()
  return read_parents

print("Math scores by parental_education\n")
parents_math = math_parents_describe(data_explore)
print(parents_math)

print("\nReading scores by parental_education \n")
parents_read = read_parents_describe(data_explore)
print(parents_read)

print("\nWriting scores by parental_education\n")
parents_write = write_parents_describe(data_explore)
print(parents_write)

Math scores by parental_education

                    count       mean        std   min   25%   50%    75%    max
parental_education                                                             
associate's degree  190.0  67.221053  14.035615  28.0  58.0  67.0  77.00  100.0
bachelor's degree   110.0  69.009091  14.511935  33.0  61.0  69.0  78.00  100.0
college             244.0  67.032787  13.934553  31.0  58.0  68.0  77.00  100.0
high school         380.0  64.402632  16.410579  13.0  53.0  65.0  76.25  100.0
master's degree      76.0  74.578947  13.958284  27.0  66.0  73.0  83.75  100.0

Reading scores by parental_education 

                    count       mean        std  ...   50%    75%    max
parental_education                               ...                    
associate's degree  190.0  70.784211  13.780623  ...  71.0  81.00  100.0
bachelor's degree   110.0  72.654545  14.784344  ...  73.5  83.00  100.0
college             244.0  69.795082  13.469636  ...  70.0  78.25  100.0


Parental eductaion level has a significant impact on the pupils' scores in all three subjects. A larger number of pupils' parents (38%) have high school education. While only 7.6% of the parents have master's degree. Pupils whose parents have master's degree score on avaerage 10 points more than pupils whose parents have finished high school in all 3 subjects. Pupils with parents finishing just high school seem to get the lowest score in all 3 subjects.